首頁 - Marquee、精選攝影師、常見問答、footer

dogwantfly opened the pull request:
https://github.com/dogwantfly/photograph-booking/issues/3#issue-1995193564

有加入全域樣式設定，若有想調整的部分可以直接修改或留言～感謝
- 參考 [Using Material UI with a custom theme](https://mui.com/material-ui/guides/next-js-app-router/#using-material-ui-with-a-custom-theme) 新增 ThemeRegistry 元件
- 將設計稿的文字、顏色樣式新增在 app/lib/theme.js

dogwantfly committed:
feat: 🎸 新增攝影師卡片切版

dogwantfly committed:
feat: 🎸 新增 Footer、常見問答切版

dogwantfly committed:
feat: 🎸 調整常見問題、Footer

dogwantfly committed:
feat: 🎸 新增 Marquee

dogwantfly committed:
feat: 🎸 新增精選攝影師、常見問題切版

dogwantfly committed:
feat: 🎸 新增填充資料、動畫效果、輪播

dogwantfly committed:
feat: 🎸 新增 MUI 全域樣式

jayredk commented on public/phorographer-bg.png:
一般的圖片可以用 [cloudconvert](https://cloudconvert.com/jpeg-to-webp) 轉成 webp（可選擇檔案是 jpg 還是 png）
Note：一天好像只有 25 張免費轉換使用。如果使用其他網站要留意圖片品質的轉換率，有些網站轉格式的效率較差，圖片解析度會變差

轉完後可以用 [TinyPng](https://tinypng.com/) 壓縮

此外，若圖片太大張可以透過 [iloveimg](https://www.iloveimg.com/zh-tw/resize-image) 調整，
以兩倍圖為原則（最大不超過要使用的圖片大小的兩倍）

這裡的建議之後可以寫進共筆當作記錄（Notion、hackmd）

jayredk commented on app/globals.css:
```diff
@@ -73,20 +73,20 @@
 
 body {
-  color: rgb(var(--foreground-rgb));
+  color: rgb(var(--background-start-rgb));
```
這個地方應該可以直接寫在 layout 的 body 標籤上，透過 style 屬性的方式
然後這支 CSS 就可以移除了

dogwantfly replied:
後來發現我需要使用到 theme 的顏色，所以我改設定在 Footer.js [這段](https://github.com/dogwantfly/photograph-booking/blob/feature/photographer-card/app/components/Footer.js#L34)
```
<Box
  sx={(theme) => ({
    backgroundColor: theme.palette.grey[900]
  })}
>
  ...
</Box>
```

jayredk commented on app/lib/theme.js:
```diff
@@ -0,0 +1,125 @@
+const oswald = Oswald({
+  subsets: ['latin'],
+  weight: ['400', '700'],
+});
```
Oswald 在這可以移除，改成按需匯入
由於字型會按順序套用，目前的順序是 Noto sans TC > Inter > Oswald，
又由於 Inter 和 Oswald 都是英數類的字型，所以一定會先套用 Inter

jayredk commented on app/lib/theme.js:
```diff
@@ -0,0 +1,125 @@
+export const primary_black = '#1F1300';
+export const pink = '#EA507B';
+export const skin = '#F1E9E1';
+export const macaron = 'rgba(255, 245, 211, 1)';
```
這裡的顏色可以直接寫進 theme 裡，不用匯出

jayredk commented on app/components/Marquee.js:
```diff
@@ -0,0 +1,110 @@
+        overflowX: 'hidden',
+        gap: 4,
+        alignItems: 'center',
+        backgroundColor: theme.macaron,
```
這裡的寫法可以改成
```
<Box
  sx={(theme) => ({
    backgroundColor: theme.palette.macaron.main
  })}
/>
```
就不需要匯入 theme 了

jayredk commented on app/components/Marquee.js:
```diff
@@ -0,0 +1,110 @@
+        gap: 4,
+        alignItems: 'center',
+        backgroundColor: theme.macaron,
+        paddingY: 2,
```
paddingY 可以簡寫成 py，其他例子如：px、pl、pr 等等，margin 也同樣適用

jayredk commented on app/components/Marquee.js:
```diff
@@ -0,0 +1,110 @@
+          flexWrap: 'nowrap',
+          gap: 4,
+          alignItems: 'center',
+          transform: 'translateX(-100px)',
```
這行也太神奇
我查過 [sx props](https://mui.com/system/properties/)
裡面沒有 transform，結果竟然直接寫有效果
印象中寫好幾次都沒效果我就改用 style 寫了

在猜 import Box from '@mui/material/Box' 和 import Box from '@mui/system/Box'; 可能是不同東西
要再了解一下

dogwantfly replied:
我看 sx props 那頁前面有提到 `You can specify any valid CSS using this prop, as well as many theme-aware properties that are unique to MUI System.`，所以除了說明有列出的那些屬性，應該 CSS 都可以（？

jayredk commented on app/components/FAQ.js:
```diff
@@ -0,0 +1,82 @@
+          onChange={handleChange(`faq${qa.id}`)}
+        >
+          <AccordionSummary
+            expandIcon={ expanded.includes(`faq${qa.id}`) ? <RemoveIcon htmlColor={theme.secondary}/>: <AddIcon htmlColor="black"/> }
```
RemoveIcon 的地方可以改成：
```
<RemoveIcon sx={(theme) => ({
  color: theme.palette.secondary.main
})}/>
```
```suggestion
            expandIcon={ expanded.includes(`faq${qa.id}`) ? <RemoveIcon sx={(theme) => ({
  color: theme.palette.secondary.main
})}/>: <AddIcon htmlColor="black"/> }
```

dogwantfly replied:
我後來改成
```
<RemoveIcon color="secondary" htmlColor="secondary"/>
```
好像也 ok~

jayredk commented on app/components/FAQ.js:
```diff
@@ -0,0 +1,82 @@
+
+const theme = {
+  secondary: 'rgba(234, 80, 123, 1)'
+}
```
這行可以移除，可以改透過下面的方式套用主題顏色
```
<RemoveIcon sx={(theme) => ({
  color: theme.palette.secondary.main
})}/>
```

jayredk commented on app/components/FAQ.js:
```diff
@@ -0,0 +1,82 @@
+              fontWeight: 700,
+              marginRight: 1,
+              lineHeight: 1.2
+            }}>Q.</Typography>
```
```suggestion
            <Typography sx={{
              fontSize: 18,
              fontWeight: 700,
              marginRight: 1,
              lineHeight: 1.2
            }}
            >
              Q.
            </Typography>
```

jayredk commented on app/components/FAQ.js:
```diff
@@ -0,0 +1,82 @@
+              marginRight: 1,
+              color: theme.secondary,
+              lineHeight: 1.2
+            }}>A.</Typography>
```
```suggestion
          <Typography sx={{
              fontSize: 18,
              fontWeight: 700,
              marginRight: 1,
              color: theme.secondary,
              lineHeight: 1.2
            }}>
              A.
            </Typography>
```

jayredk commented on app/components/Footer.js:
```diff
@@ -0,0 +1,87 @@
+  Copyright © 2023 隨旅拍
+</Typography>
+      </Container>
+  );
```
縮排可以調整一下，
之後看要不要加個 prettier，[參考老師的文章](https://www.casper.tw/development/2021/04/11/vscode-eslint-prettier/)

```
  return (
    <Container
      component="footer"
      sx={{
        position: 'relative',
        color: 'white',
        textAlign: 'center',
        marginY: {'xs': 3, 'md': 6}
      }}
    >
      <Box sx={{
        position: 'absolute',
        left: '24px',
        right:'24px',
        top: 0,
        bottom: 0,
        display: 'flex',
        justifyContent: 'space-between'
      }}>
        <Box sx={{
          display: 'flex',
          flexDirection: 'column',
          height: '100%'
        }}>
        <Box sx={{
          border: '1px solid white',
          borderRight: 0,
          borderBottom: 0,
          width: 12,
          height: 12,
          marginBottom: 'auto'
        }}/>
        <Box sx={{
          border: '1px solid white',
          borderRight: 0,
          borderTop: 0,
          width: 12,
          height: 12
        }}/>
        </Box>
        <Box sx={{
          display: 'flex',
          flexDirection: 'column',
          height: '100%'
        }}>
          <Box sx={{
            border: '1px solid white',
            borderLeft: 0,
            borderBottom: 0,
            width: 12,
            height: 12,
            marginBottom: 'auto'
          }}/>
          <Box sx={{
            border: '1px solid white',
            borderLeft: 0,
            borderTop: 0,
            width: 12,
            height: 12
          }}/>
        </Box>
      </Box>
      <Box sx={{
        width: {
          xs: '120px',
          sm: '240px'
        },
        margin: '0 auto'
      }}>
        <Image 
          alt="隨旅拍"
          src="/logo_light.svg"
          width={240}
          height={64}
          sizes="(min-width: 768px) 240px 64px, 120px 32px"
          style={{ maxWidth: '100%' }}
        />
      </Box>

      <Typography sx={{
        fontSize: 14,
      }}>
        Copyright © 2023 隨旅拍
      </Typography>
    </Container>
  );
```

另外，第 75 行圖片的斷點是使用自訂的，看要不要統一使用 MUI 的斷點
我是寫成這樣：
```
const theme = useTheme();
const matchesMd = useMediaQuery(theme.breakpoints.up('md'));
const matchesSm = useMediaQuery(theme.breakpoints.up('sm'));

const getImgWidth = () => {
    if (matchesMd) {
      return 236;
    } else if (matchesSm) {
      return 150;
    } else {
      return 104;
    }
  };
  
<Image 
  src={logo}
  width={getImgWidth()}
  style={{
    maxWidth: '100%',
    height: 'auto',
  }}
  alt='logo'
/>
```

jayredk commented on app/components/ThemeRegistry.js:
```diff
@@ -0,0 +1,63 @@
+'use client';
+import createCache from '@emotion/cache';
+import { useServerInsertedHTML } from 'next/navigation';
+import { CacheProvider } from '@emotion/react';
```
cache 的部分可以再拆成一支檔案，看起來比較簡潔一些
[參考連結](https://github.com/dogwantfly/photograph-booking/blob/feat/banner/app/components/ThemeRegistry/EmotionCache.js)

jayredk commented on app/components/ThemeRegistry.js:
```diff
@@ -0,0 +1,63 @@
+import { ThemeProvider } from '@mui/material/styles';
+import { useState } from 'react';
+import CssBaseline from '@mui/material/CssBaseline';
+import theme from '@/app/lib/theme'
```
theme 的部分也可以改移到 ThemeRegistry 資料夾下
因為主要用到 theme 的就是 ThemeRegistry.js 這個元件
[參考連結](https://github.com/dogwantfly/photograph-booking/tree/feat/banner/app/components/ThemeRegistry)

jayredk commented on app/page.js:
```diff
@@ -1,95 +1,140 @@
+          position: 'relative',
+          overflowX: 'hidden',
+        }}
+      >
```
常見問題的區塊可以在 components 新增一個資料夾，例如叫 Faq
裡面可以放跟常見問題有關的元件
最後透過 index.js 匯出
再匯入到 page.js

jayredk commented on app/components/PhotographerSection.js:
```diff
@@ -0,0 +1,234 @@
+import { Swiper, SwiperSlide } from 'swiper/react';
+import { FreeMode } from 'swiper/modules';
+import 'swiper/css';
+import 'swiper/css/free-mode';
```
import 順序可以稍微調整一下，相關性的可以放一起，順序原則上大概是：react 相關 > 第三方套件 > 元件 > Icon + 字型 > 圖片匯入

```suggestion
import 'swiper/css';
import 'swiper/css/free-mode';

import { useRef, useLayoutEffect } from 'react';
import { Swiper, SwiperSlide } from 'swiper/react';
import { FreeMode } from 'swiper/modules';
import Container from '@mui/material/Container';
import Typography from '@mui/material/Typography';
import Box from '@mui/material/Box';
import Link from '@mui/material/Link';

import { gsap } from 'gsap';
import { ScrollTrigger } from 'gsap/ScrollTrigger';

import Marquee from '@/app/components/Marquee';
import PhotographerCard from '@/app/components/PhotographerCard';
import { photographers } from '@/app/lib/placeholder-data';

import { Noto_Sans_TC } from 'next/font/google';
import EastIcon from '@mui/icons-material/East';
import photographerBg from '../../public/phorographer-bg.png';

```

jayredk commented on app/components/PhotographerSection.js:
```diff
@@ -0,0 +1,234 @@
+                  slidesPerView: 4,
+                  spaceBetween: 30,
+                },
+              }}
```
這段的 0.00、0.75、1.00 分別代表什麼斷點
看起來不是用 MUI 的斷點

jayredk commented on app/components/PhotographerCard.js:
```diff
@@ -0,0 +1,203 @@
+          right: 0,
+          zIndex: 1,
+        }}
+      >
```
```suggestion
      <CardContent
        sx={{
          padding: 0,
          position: 'absolute',
          top: 0,
          bottom: 0,
          left: 0,
          right: 0,
          zIndex: 1,
        }}
      >
```

jayredk commented on app/components/PhotographerCard.js:
```diff
@@ -0,0 +1,203 @@
+          width: '100%',
+          height: '100%'
+        }}
+      /> 
```
```suggestion
        <CardActionArea
          component="a"
          href="/"
          sx={{
            width: '100%',
            height: '100%'
          }}
        /> 
```

jayredk commented on app/components/PhotographerSection.js:
```diff
@@ -0,0 +1,234 @@
+            >
+              瀏覽更多
+              <EastIcon sx={{ color: 'white', marginLeft: 1.5 }} />
+            </Link>
```
可以補一下 hover 效果
<img width="497" alt="圖片" src="https://github.com/dogwantfly/photograph-booking/assets/47918755/2c9a45fc-82cb-4699-ba28-ed216fd74071">

jayredk commented on app/components/PhotographerCard.js:
```diff
@@ -0,0 +1,203 @@
+      sx={{
+        position: 'relative',
+      }}
+    >
```
```suggestion
    <Card
      sx={{
        position: 'relative',
        borderRadius: '0.75rem'
      }}
    >
```

jayredk commented on app/components/PhotographerCard.js:
```diff
@@ -0,0 +1,203 @@
+        >
+          <Typography
+            sx={{
+              color: 'rgba(31, 19, 0, 1)',
```
這裡設計稿用黑色，總覺得哪裡怪怪的，感覺白色好一些（？

```suggestion
              color: 'primary.main',
```

dogwantfly replied:
那我這邊先改為白色~

jayredk commented on app/components/PhotographerCard.js:
```diff
@@ -0,0 +1,203 @@
+            sx={{
+              fontSize: 14,
+              fontWeight: 700
+            }}
```
```suggestion
            sx={{
              fontSize: 14,
              fontWeight: 700,
              letterSpacing: '0.0175rem'
            }}
```

jayredk commented on app/components/PhotographerCard.js:
```diff
@@ -0,0 +1,203 @@
+          </Stack>
+          <Typography className={noto_sans_tc.className}>
+            NT$ {price.toLocaleString("en-US")} / 小時
+          </Typography>
```
這裡的 NT$ {price} 的字型是 Inter，只有 '/ 小時' 是 Noto sans TC

jayredk commented on app/components/PhotographerCard.js:
```diff
@@ -0,0 +1,203 @@
+            '&::-webkit-scrollbar': {
+            display: 'none'
+          }
+          }}>
```
這裡的滾動條會動
::-webkit-scrollbar 只有 chrome 有效，且不能拉（等於 overflow:hidden 的效果）

這裡可以先設 overflowX: 'hidden'
之後再看看有什麼其他方法可以讓使用者也能在沒有滾動條的情況下滾動標籤
目前情況滾動條跟 swiper 的功能也有重疊到

jayredk reviewed (commented):
從發 PR 到 Review PR 的過程可能要再調整
PR 的東西有點多，可能會不知道從何下手、看到多細
之後可以嘗試一個區塊發一個 PR、有動畫、套件的部分另外發
一個 PR 以 2-300 行左右為限，且只做某一個區塊、功能
[參考文章](https://www.cythilya.tw/2020/12/10/google-how-to-get-through-code-review/#%E5%B0%8F%E7%9A%84-cl)
[參考影片](https://www.youtube.com/watch?v=Zg0fcg016ko)

[Review 小技巧](https://www.facebook.com/groups/git.tw/posts/2990438154338166/)

Review 的過程大概是：
```
git fetch origin
git checkout <branch-name>
git pull origin <branch-name>
```
先將分支拉到本地端，接著看看功能、有沒有破版

接著點開 code
大概先看一些 config 設定、再來就是排版、樣式寫法
大致對了一些設計稿，有發現幾個地方樣式沒有設定到
RWD 的部分主要只看了是否有正常變換版型，其餘樣式沒有細看

排版的部分可以考慮加入 prettier，[參考老師文章](https://www.casper.tw/development/2021/04/11/vscode-eslint-prettier/)
斷點的統一、樣式的寫法、圖片的優化等等之後應該可以寫一份文件當作統一標準

dogwantfly committed:
chore: 🤖 重安裝 gasp 套件

dogwantfly committed:
perf: ⚡️ 調整圖片優化

dogwantfly committed:
chore: 🤖 移除圖片

dogwantfly committed:
refactor: 💡 移除 theme.js Oswald 字型

dogwantfly committed:
refactor: 💡 移除 globals.css

dogwantfly committed:
refactor: 💡 修改 theme 顏色套用

dogwantfly committed:
refactor: 💡 調整 padding、margin 寫法

dogwantfly committed:
refactor: 💡 調整 padding、margin 寫法

dogwantfly committed:
refactor: 💡 調整程式碼排版

dogwantfly committed:
refactor: 💡 調整 footer Image 斷點，統一使用 MUI 斷點

dogwantfly committed:
refactor: 💡 修改 ThemeRegistry

參考範例新增 ThemeRegistry 資料夾、拆出 EmotionCache.js、index.js、將 theme.js 移至
ThemeRegistry 資料夾

dogwantfly committed:
refactor: 💡 調整常見問答區塊

調整 ul 樣式、新增 Faq 元件資料夾

dogwantfly committed:
style: 💄 調整 import 順序

dogwantfly committed:
refactor: 💡 調整 swiper 斷點設定

dogwantfly committed:
refactor: 💡 加入 Link hover 效果

dogwantfly committed:
refactor: 💡 調整攝影師卡片細節
